src/main/java/org/elephant/sam/SAMDetectionTask.java

```diff
@@ -0,0 +1,385 @@
+
+    private final ImageData<BufferedImage> imageData;
+    private ImageServer<BufferedImage> renderedServer;
+
+    /**
+     * The field of view visible within the viewer at the time the detection task was created.
+     */
+    private RegionRequest viewerRegion;
+
+    private double padScale = 2.0;
+
+    private final List<PathObject> foregroundObjects;
+    private final List<PathObject> backgroundObjects;
+
+    private final boolean setRandomColor;
+    private final boolean setName;
+    private final boolean keepPromptObjects;
+    private final SAMOutput outputType;
+
+    private final String serverURL;
+
+    private final SAMModel model;
+
+    private SAMDetectionTask(Builder builder) {
+        this.serverURL = builder.serverURL;
+        Objects.requireNonNull(serverURL, "Server must not be null!");
+
+        this.model = builder.model;
+        Objects.requireNonNull(model, "Model must not be null!");
+
+        QuPathViewer viewer = builder.viewer;
+        Objects.requireNonNull(builder, "Viewer must not be null!");
+
+        this.imageData = viewer.getImageData();
+        Objects.requireNonNull(imageData, "ImageData must not be null!");
+        if (builder.server != null) {
+            if (builder.server.isRGB())
+                this.renderedServer = builder.server;
+            else
+                logger.warn("Cannot use non-RGB image server for detection!");
+        }
+        if (this.renderedServer == null) {
+            this.renderedServer = Utils.createRenderedServer(viewer);
+        }
+
+        // Find the region and downsample currently used within the viewer
+        ImageRegion region = AwtTools.getImageRegion(viewer.getDisplayedRegionShape(), viewer.getZPosition(), viewer.getTPosition());
+        this.viewerRegion = RegionRequest.createInstance(renderedServer.getPath(), viewer.getDownsampleFactor(), region);
+        this.viewerRegion = viewerRegion.intersect2D(0, 0, renderedServer.getWidth(), renderedServer.getHeight());
+
+        this.foregroundObjects = new ArrayList<>(builder.foregroundObjects);
+        this.backgroundObjects = new ArrayList<>(builder.backgroundObjects);
+
+        this.keepPromptObjects = builder.keepPromptObjects;
+        this.outputType = builder.outputType;
+        this.setName = builder.setName;
+        this.setRandomColor = builder.setRandomColor;
+    }
+
+    @Override
+    protected List<PathObject> call() throws Exception {
+        try {
+            List<PathObject> detected = detectObjects();
+            if (!detected.isEmpty()) {
+                PathObjectHierarchy hierarchy = imageData.getHierarchy();
+                if (!keepPromptObjects) {
+                    // Remove prompt objects in one step
+                    hierarchy.getSelectionModel().clearSelection();
+                    hierarchy.removeObjects(foregroundObjects, true);
+//                    List<PathObject> toRemove = new ArrayList<>(foregroundObjects);
```

Comment:
Guessing from the comment-out lines here, is it better to keep the background objects?

Reply:
I thought so, but I'm not sure... curious as to your experience with it. 

I expected that I would draw multiple background objects for things that shouldn't be segmented, and then one click per object that should be segmented. Then the background points should be kept - because the user could want the background to be 'remembered' while live annotating.

However I found if I added many background objects (e.g. 5-10), my foreground predictions were **terrible**. As in, the foreground prediction was often nowhere near by foreground point.

I don't know if something in my code was the problem, but there are some hints [here](https://github.com/search?q=repo%3Afacebookresearch%2Fsegment-anything+multiple+points&type=issues) that others have seem similar behavior when using multiple points.

With that in mind, I think a more likely way to use background points would be to click at most one or two background points for each foreground point. In this case, I think they *should* be removed at the same time (and the part of my code that is commented-out should be restored). Because it doesn't make sense to keep multiple background points before annotating the foreground if the background just destroys the result.